docs/tsm/TSI_CARDINALITY_PROPOSAL.md

```diff
@@ -0,0 +1,254 @@
+
+#### TSI file format
+
+We could extend the TSI file format to include a sketches block for series and measurements, which could be for example at the beginning of the index.
+We would also need to add tombstone counters for series and measurements, which I have not added to this diagram, but would need to be considered in #7174.
+
+```
+╔═══════Inverted Index═══════╗
+║ ┌────────────────────────┐ ║
+║ │                        │ ║
+║ │        Sketches        │ ║
+║ │                        │ ║
+║ └────────────────────────┘ ║
+║ ┌────────────────────────┐ ║
+║ │                        │ ║
+║ │   Series Dictionary    │ ║
+║ │                        │ ║
+║ └────────────────────────┘ ║
+║ ┌────────Tag Set─────────┐ ║
+║ │┌──────────────────────┐│ ║
+║ ││   Tag Values Block   ││ ║
+║ │├──────────────────────┤│ ║
+║ ││   Tag Values Block   ││ ║
+║ │├──────────────────────┤│ ║
+║ ││    Tag Keys Block    ││ ║
+║ │└──────────────────────┘│ ║
+║ └────────────────────────┘ ║
+║ ┌────────Tag Set─────────┐ ║
+║ │┌──────────────────────┐│ ║
+║ ││   Tag Values Block   ││ ║
+║ │├──────────────────────┤│ ║
+║ ││   Tag Values Block   ││ ║
+║ │├──────────────────────┤│ ║
+║ ││   Tag Values Block   ││ ║
+║ │├──────────────────────┤│ ║
+║ ││    Tag Keys Block    ││ ║
+║ │└──────────────────────┘│ ║
+║ └────────────────────────┘ ║
+║  ┌──────────────────────┐  ║
+║  │  Measurements Block  │  ║
+║  └──────────────────────┘  ║
+╚════════════════════════════╝
+```
+
+The sketch block itself would contain a value indicating the number of sketches in the block.
+Each sketch block would then contain the size of the sketch data, and the sketch itself.
+The sketch may be a binary blob, perhaps generated by Go's `gob` package, or it may be a custom binary format of data comprising the HLL algorithm state and sketch.
+The exact format will be decided during implementation.
+
+```
+╔═══════════Sketches═══════════╗
+║ ┌──────────────────────────┐ ║
+║ │   Sketch Count <uint32>  │ ║
+║ └──────────────────────────┘ ║
+║ ┌────── Series Sketch ─────┐ ║
+║ │┌────────────────────────┐│ ║
+║ ││  len(Sketch) <uint32>  ││ ║
+║ │└────────────────────────┘│ ║
+║ │┌────────────────────────┐│ ║
+║ ││    Sketch <byte...>    ││ ║
+║ │└────────────────────────┘│ ║
+║ └──────────────────────────┘ ║
+║ ┌────Measurements Sketch───┐ ║
+║ │┌────────────────────────┐│ ║
+║ ││  len(Sketch) <uint32>  ││ ║
+║ │└────────────────────────┘│ ║
+║ │┌────────────────────────┐│ ║
+║ ││    Sketch <byte...>    ││ ║
+║ │└────────────────────────┘│ ║
+║ └──────────────────────────┘ ║
+╚══════════════════════════════╝
+```
+
+### WAL file format
+
+Each WAL file would also have a Sketches block, which is the same format as the block in the TSI index.
+These sketches however only contain cardinality estimations for whatever is in the WAL file.
+
+```
+╔═════════════WAL══════════════╗
+║ ┌──────────────────────────┐ ║
+║ │                          │ ║
+║ │         Sketches         │ ║
```

Comment:
Yep. See earlier comments on this PR with Ben. I think we could just keep an in-memory WAL sketch for each WAL file, and then just re-build them on startup via a scan of each WAL file. Since WAL files are relatively small this shouldn't be too bad.